Extra sampling of record function events

ilia-cher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#48289 Extra sampling of record function events**

Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

Commit:
Record top level operators in RecordFunction

Summary:
This PR changes default behavior of RecordFunction to only record
top level operators in dispatcher, note - this does not change profiler
use case (we continue recording all the ops for profiler)

Test Plan:
benchmark results: tbd

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -371,28 +371,31 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
-        guard.before(op, stack, seq_num);
-      } else {
-        guard.before(op, seq_num);
+  if (!at::isRecordingTopLevelOnly() || at::isTopLevelRecordFunction()) {
```
this accesses TLS twice, if you put them in a single function, you can do it once

Commit:
Update on "Record top level operators in RecordFunction"


Summary:
This PR changes default behavior of RecordFunction to only record
top level operators in dispatcher, note - this does not change profiler
use case (we continue recording all the ops for profiler)

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

swolchok commented on aten/src/ATen/record_function.cpp:
```diff
@@ -428,4 +428,55 @@ void RecordFunction::end() {
+//
+namespace {
+std::atomic<int> global_record_all_functions_ {0};
+thread_local int tls_record_all_functions_ = 0;
```
you can save on TLS accesses by grouping these two into a struct

ezyang replied:
Yeah, this was also something that popped out to me when I saw this patch

swolchok commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -429,20 +432,24 @@ inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const
-        guard.before(op, *stack, seq_num);
-      } else {
-        guard.before(op, seq_num);
+  if (at::shouldRunRecordFunction()) {
```
is performance of the case where we *do* need to record all functions important? This should be a regression in that case because of (sadly) the extra function call overhead.

ilia-cher replied:
it's important for specific use cases (mainly, profiler), that use this; but normally in production profiler won't be enabled; existing production observers are already heavy sampled; in this pr we move sampling partially up, into Dispatcher.h, to avoid creating RecordFunction object altogether

swolchok commented on aten/src/ATen/record_function.cpp:
```diff
@@ -428,4 +428,55 @@ void RecordFunction::end() {
+}
+
+bool shouldRunRecordFunction() {
+  if (global_record_all_functions_ > 0) {
```
if we don't care about a little slop, this can be `global_record_all_functions_.load(std::memory_order_relaxed)` and will perform better on non-x86 architectures. similarly, I think the increment/decrement can use relaxed as well in that case.

swolchok reviewed (commented):
Should the RecordFunction macro in RecordFunction.h also be gated? Seems like setExtraSamplingProbability is under-specified/under-documented.

This hugely reduces sampling, right? Is that OK? Why were we recording so much before if we didn't need to?

swolchok commented:
Does docs/source/notes/large_scale_deployments.rst need to be updated too?

ezyang commented:
It doesn't seem to me like there's enough description of the PR talking about what is being done and what the tradeoffs here are.

ilia-cher commented:
I haven't yet added enough docs - will improve

@swolchok , @ezyang 

We just split the sampling in two parts - (semi)mandatory sampling with kLowProb (which we can make fast) and, if necessary, extra sampling; so we only need to adjust the sampling rates to account for an extra sampling; the end result is that we just skip most of the RecordFunction object creation in Dispatcher.h but still are able to use sampling callbacks and non-sampled too

ilia-cher commented:
what we essentially change here is the tradeoff between a large number of RecordFunction objects with small probabilities and a smaller number of RecordFunction objects with larger probabilities

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

ilia-cher commented:
PR's updated:
no need to modify user contract, from user point of view things stay and work the same, internally we use an optimization when no high-frequency or non-sampled callbacks are used;

ready for review @ngimel @dzhulgakov

ezyang commented on aten/src/ATen/record_function.cpp:
```diff
@@ -428,4 +457,41 @@ void RecordFunction::end() {
+  TORCH_CHECK(--global_record_all_functions_ >= 0);
+}
+
+bool shouldRunRecordFunction(bool& pre_sampled) {
```
I want to complain about this mutable reference but I recently found out Google changed their C++ style guide to prefer mutable reference output parameters because they can't be null 😂. It's still pretty obscure...

ezyang replied:
I feel that the documentation explaining what is going on is more appropriately placed here, because there are multiple call sites (but you have only added a big doc at one of the call sites; what if a reader comes from another call site?)

ezyang commented:
What's the testing strategy for this PR? I see in the diff you posted benchmark numbers. But it's easy to get things wrong and, for example, completely turn off profiling, which would also speed things up but also be totally wrong...

ezyang commented on aten/src/ATen/record_function.cpp:
```diff
@@ -428,4 +457,41 @@ void RecordFunction::end() {
+  auto* rf_tls_ptr = &rf_tls_;
+  if (!rf_tls_ptr->tls_record_function_enabled_) {
+    pre_sampled = false;
+    return false;
```
If you return false here, is there any point in setting `pre_sampled`?

ilia-cher replied:
then we should say 'if function returns false, pre_sampled argument value is unchanged/uninitialized', i slightly prefer to always have a deterministic out value

ezyang commented:
What are the benchmarks after you aren't hella downsampling? If the problem is really RecordFunction overhead then naively I would expect that once you restored sampling rate to the normal amount, the performance improvement would also vanish.

ngimel commented on aten/src/ATen/record_function.cpp:
```diff
@@ -29,11 +29,9 @@ std::atomic<int64_t> defaultNodeId(-1);
 // Low probability constant
 const double kLowProb = 0.001;
-thread_local int tries_left_ = 0;
+const double kEps = 10e-9;
```
nit: 1e-8?

ngimel commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -371,28 +371,38 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
+        }
       }
     }
+    return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
```
Leave a comment that you need guard alive, so you call the kernel here?

ngimel commented on aten/src/ATen/record_function.cpp:
```diff
@@ -255,7 +266,16 @@ namespace {
+  if (pre_sampled) {
+    // RecordFunction is already pre-sampled with prob. kLowProb
+    if (sampling_prob_ > (kLowProb + kEps)) {
+      TORCH_CHECK(
```
Seems like it should be TORCH_INTERNAL_ASSERT?

ngimel commented on aten/src/ATen/record_function.cpp:
```diff
@@ -265,23 +285,31 @@ bool RecordFunctionCallback::shouldRun(RecordScope scope) const {
+    // the RecordFunction
+    sampling_prob /= kLowProb;
+  }
+  if (sampling_prob < (1.0 - kEps)) {
```
can you leave a comment why you need kEps here?

ilia-cher commented:
> What's the testing strategy for this PR? I see in the diff you posted benchmark numbers. But it's easy to get things wrong and, for example, completely turn off profiling, which would also speed things up but also be totally wrong...

not sure I understand - profiler is well covered by unit tests

ilia-cher commented:
> What are the benchmarks after you aren't hella downsampling? If the problem is really RecordFunction overhead then naively I would expect that once you restored sampling rate to the normal amount, the performance improvement would also vanish.

@ezyang - we don't change effective sampling rate, so not sure this argument applies

ilia-cher commented:
@ezyang - once again, the optimization is to avoid creating too many RecordFunction objects in case when they are active very rarely

ngimel commented:
> > What's the testing strategy for this PR? I see in the diff you posted benchmark numbers. But it's easy to get things wrong and, for example, completely turn off profiling, which would also speed things up but also be totally wrong...
> 
> not sure I understand - profiler is well covered by unit tests

@ilia-cher, agreed, regular "always on" profiling is covered in the tests, for the observers you could do similar thing to what you did in the original pre-sampling PR and make sure that profiling is triggered expected number of times.

ilia-cher commented:
> > > What's the testing strategy for this PR? I see in the diff you posted benchmark numbers. But it's easy to get things wrong and, for example, completely turn off profiling, which would also speed things up but also be totally wrong...
> > 
> > 
> > not sure I understand - profiler is well covered by unit tests
> 
> @ilia-cher, agreed, regular "always on" profiling is covered in the tests, for the observers you could do similar thing to what you did in the original pre-sampling PR and make sure that profiling is triggered expected number of times.

I see, yes I remember we had a binary to output that, will include in the test plan

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

ilia-cher commented:
addressed comments, added changes to record_function_benchmark - verified that number of calls is right https://gist.github.com/ilia-cher/2a33a207194810c6be73421db8e3f6c6

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

Commit:
Update on "Extra sampling of record function events"


Summary:
Adding extra sampling step when dispatching RecordFunction.

Test Plan:
benchmark results: tbd

Differential Revision: [D25111515](https://our.internmc.facebook.com/intern/diff/D25111515)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken pytorch_linux_xenial_cuda10_2_cudnn7_py3_slow_test. 

```
test_collect_cpp_callgrind - TestBenchmarkUtils
benchmark_utils/test_benchmark_utils.py

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 542, in wrapper
    fn(*args, **kwargs)
  File "benchmark_utils/test_benchmark_utils.py", line 530, in test_collect_cpp_callgrind
    min(counts), max(counts), "C++ Callgrind should be deterministic")
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1180, in assertEqual
    self.assertTrue(result, msg=msg)
AssertionError: False is not true : C++ Callgrind should be deterministic
```
